In [69]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [70]:
df = pd.read_csv('weekly_driving_profiles.csv')

In [ ]:
cols = ['clear_weather','weather_wind_speed_mean','weather_visibility_mean','forward_collision', 'distracted_driver','too_close_distance','lane_departure','driver_making_calls','driver_smoking','fatigue_driving']

df = df.drop(columns=cols)

In [71]:
import os
import wandb # для логирования

import numpy as np
import random
from tqdm import *

import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim # для оптимизаторов
from torchvision import datasets # для данных
import torchvision.transforms as transforms # для преобразований тензоров


In [72]:
class CFG:

# Задаем параметры нашего эксперимента
  feature_cols = ['engine_capacity', 'road_quality_moderate', 'slope_flat',
                'motorway', 'rural', 'more_than_one_lane', 'congested',
                'speed_limit_mean', 'weather_temperature_mean', 'total_distance',
                'sum_roundabout', 'sum_traffic_signal', 'sum_stop_sign',
                'sum_yield_sign', 'sum_pedestrian_crossing_sign',
                'sum_animal_crossing_sign', 'speeding_serious',
                'harsh_acceleration', 'harsh_braking']
  input_dim = len(feature_cols)
  hidden_dims = [32, 16, 8]
  dropout_rate = 0.2
  learning_rate = 0.001
  num_epochs = 60

hidden_dims = [32, 16, 8] - была совершена проверка (сравнение) разного количества нейронов на 1 слое

Каждый следующий шаг (слой) обобщает информацию (уменьшение нейронов), находии общие зависимости

Adam оптимизатор (по умолчанию) работает лучше всего с lr=0.001 - базовый выбор

60 эпох - практическая проверка. Лучшее обучение на более ранней эпохе, дальше уже переобучение начинается



In [73]:
from torch.utils.data import TensorDataset, DataLoader
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

feature_cols = ['engine_capacity', 'road_quality_moderate', 'slope_flat',
                'motorway', 'rural', 'more_than_one_lane', 'congested',
                'speed_limit_mean', 'weather_temperature_mean', 'total_distance',
                'sum_roundabout', 'sum_traffic_signal', 'sum_stop_sign',
                'sum_yield_sign', 'sum_pedestrian_crossing_sign',
                'sum_animal_crossing_sign', 'speeding_serious',
                'harsh_acceleration', 'harsh_braking']
target_col = 'claims_count'

unique_drivers = df['driver_id'].unique()

#примерно 70% обучение 30% тест
train_drivers, test_drivers = train_test_split(unique_drivers, test_size=0.3, random_state=30)

X_train = df[df['driver_id'].isin(train_drivers)][feature_cols]
X_test = df[df['driver_id'].isin(test_drivers)][feature_cols]
y_train = df[df['driver_id'].isin(train_drivers)][target_col]
y_test = df[df['driver_id'].isin(test_drivers)][target_col]

print("Количество водителей в тренировке: ",len(train_drivers), "Количество строк: ",len(X_train))
print("Количество водителей в тесте: ",len(test_drivers), "Количество строк: ",len(X_test))


#масштабирование признаков
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


X_train_t = torch.tensor(X_train, dtype=torch.float32)
y_train_t = torch.tensor(y_train.values, dtype=torch.float32).reshape(-1, 1)
X_test_t = torch.tensor(X_test, dtype=torch.float32)
y_test_t = torch.tensor(y_test.values, dtype=torch.float32).reshape(-1, 1)


train_data = TensorDataset(X_train_t, y_train_t)
test_data = TensorDataset(X_test_t, y_test_t)

#рекомендуют для датасета 10к-100к батчсайз 64
batch_size = 64

train_loader = DataLoader(train_data, batch_size=batch_size, shuffle=True)
test_loader = DataLoader(test_data, batch_size=batch_size, shuffle=False)


Количество водителей в тренировке:  247 Количество строк:  8465
Количество водителей в тесте:  107 Количество строк:  4063


In [74]:
def seed_everything(seed):
    random.seed(seed) # фиксируем генератор случайных чисел
    os.environ['PYTHONHASHSEED'] = str(seed) # фиксируем заполнения хешей
    np.random.seed(seed) # фиксируем генератор случайных чисел numpy
    torch.manual_seed(seed) # фиксируем генератор случайных чисел pytorch
    torch.cuda.manual_seed(seed) # фиксируем генератор случайных чисел для GPU
    #torch.backends.cudnn.deterministic = True # выбираем только детерминированные алгоритмы (для сверток)
    #torch.backends.cudnn.benchmark = False # фиксируем алгоритм вычисления сверток
seed_everything(30)

In [75]:
class Regression(nn.Module): # наследуемся от класса nn.Module
    def __init__(self):
        super(Regression,self).__init__()
        # организуем 3 скрытых слоя
        hidden_1 =  CFG.hidden_dims[0]
        hidden_2 = CFG.hidden_dims[1]
        hidden_3 = CFG.hidden_dims[2]
        #
        input_dim = 19
        self.net = torch.nn.Sequential(
                      torch.nn.Linear(input_dim, hidden_1),
                      torch.nn.ReLU(),
                      torch.nn.Linear(hidden_1, hidden_2),
                      torch.nn.ReLU(),
                      torch.nn.Linear(hidden_2, hidden_3),
                      torch.nn.ReLU(),
                      torch.nn.Linear(hidden_3, 1),
                    )

    def forward(self,x):
        x = torch.exp(self.net(x))
        return x

3 слоя Linear - базовая архитектура

ReLU() - базовая функция активации

torch.exp(self.net(x)) - экспонента для положительного результата

In [76]:
model = Regression()
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = model.to(device) # переводим модель на GPU, не получилось, поэтому CPU
print(model) # посмотрим на нашу модель


Regression(
  (net): Sequential(
    (0): Linear(in_features=19, out_features=32, bias=True)
    (1): ReLU()
    (2): Linear(in_features=32, out_features=16, bias=True)
    (3): ReLU()
    (4): Linear(in_features=16, out_features=8, bias=True)
    (5): ReLU()
    (6): Linear(in_features=8, out_features=1, bias=True)
  )
)


In [77]:
# функция потерь
criterion = nn.MSELoss()

optimizer = torch.optim.Adam(model.parameters(),lr = 0.001)

Для задачи регрессии, где нам важно больше штрафовать за большие ошибки, чем за маленькие хорошо подходит MSE

Adam - базовый хороший оптимизатор


In [78]:
# функция обучения модели
def train(model, device, train_loader, optimizer, criterion, epoch):
    model.train() # обязательно переводим в режим обучения
    train_loss_sum = 0


    n_ex = len(train_loader)

    for batch_idx, (data, target) in tqdm(enumerate(train_loader), total=n_ex):
        data, target = data.to(device), target.to(device) # переводим картинки и таргеты на GPU
        # обнуляем градиенты!
        optimizer.zero_grad()
        # прямой проход
        output = model(data)

        train_loss = criterion(output, target) # считаем значение функции потерь
        # обратный проход
        train_loss.backward()
        # делаем градиентный шаг оптимизатором
        optimizer.step()
        # считаем метрики и лосс
        train_loss_sum += train_loss.item() * data.size(0)

    tqdm.write('\nTrain set: Average loss: {:.4f}'.format(
        train_loss_sum / len(train_loader.dataset)))



In [79]:
# функция тестирования
def test(model, device, test_loader, criterion):
    model.eval() # переводем модель в режим инференса
    test_loss_sum = 0

    all_predictions = []
    all_targets = []

    # показываем, что обученич нет и градиенты не обновляются
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss = criterion(output, target) # считаем значение функции потерь
            test_loss_sum += test_loss.item() * data.size(0)

            predictions = output.cpu().numpy().flatten()
            targets = target.cpu().numpy().flatten()

            all_predictions.extend(predictions.tolist())
            all_targets.extend(targets.tolist())


            # считаем метрики
    tqdm.write('Test set: Average loss: {:.4f}'.format(
       test_loss_sum / len(test_loader.dataset)))
    return test_loss_sum / len(test_loader.dataset)



In [80]:
# основная функция для экспериментов
def main(model):

    seed_everything(30)

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu") # выделили устройство
    model = model.to(device)

    losses = []
    for epoch in range(1, CFG.num_epochs + 1): # цикл на эпохи
        print('\nEpoch:', epoch)
        train(model, device, train_loader, optimizer, criterion, epoch)
        losses.append(test(model, device, test_loader, criterion))
    print('Training is ended!')
    best_epoch = np.argmin(losses)+1
    print()
    print('BEST EPOCH: ', best_epoch, 'with Loss: ',min(losses))


In [81]:
main(model)



Epoch: 1


100%|██████████| 133/133 [00:00<00:00, 353.65it/s]



Train set: Average loss: 0.9617
Test set: Average loss: 0.7692

Epoch: 2


100%|██████████| 133/133 [00:00<00:00, 476.40it/s]



Train set: Average loss: 0.7187
Test set: Average loss: 0.6316

Epoch: 3


100%|██████████| 133/133 [00:00<00:00, 491.45it/s]



Train set: Average loss: 0.5947
Test set: Average loss: 0.5313

Epoch: 4


100%|██████████| 133/133 [00:00<00:00, 459.06it/s]



Train set: Average loss: 0.5036
Test set: Average loss: 0.4572

Epoch: 5


100%|██████████| 133/133 [00:00<00:00, 465.31it/s]



Train set: Average loss: 0.4352
Test set: Average loss: 0.4014

Epoch: 6


100%|██████████| 133/133 [00:00<00:00, 480.08it/s]



Train set: Average loss: 0.3829
Test set: Average loss: 0.3586

Epoch: 7


100%|██████████| 133/133 [00:00<00:00, 356.79it/s]



Train set: Average loss: 0.3420
Test set: Average loss: 0.3252

Epoch: 8


100%|██████████| 133/133 [00:00<00:00, 331.09it/s]



Train set: Average loss: 0.3097
Test set: Average loss: 0.2985

Epoch: 9


100%|██████████| 133/133 [00:00<00:00, 326.69it/s]



Train set: Average loss: 0.2836
Test set: Average loss: 0.2772

Epoch: 10


100%|██████████| 133/133 [00:00<00:00, 330.39it/s]



Train set: Average loss: 0.2624
Test set: Average loss: 0.2599

Epoch: 11


100%|██████████| 133/133 [00:00<00:00, 290.05it/s]



Train set: Average loss: 0.2450
Test set: Average loss: 0.2456

Epoch: 12


100%|██████████| 133/133 [00:00<00:00, 294.98it/s]



Train set: Average loss: 0.2305
Test set: Average loss: 0.2339

Epoch: 13


100%|██████████| 133/133 [00:00<00:00, 338.69it/s]



Train set: Average loss: 0.2184
Test set: Average loss: 0.2241

Epoch: 14


100%|██████████| 133/133 [00:00<00:00, 449.84it/s]



Train set: Average loss: 0.2081
Test set: Average loss: 0.2158

Epoch: 15


100%|██████████| 133/133 [00:00<00:00, 478.86it/s]



Train set: Average loss: 0.1994
Test set: Average loss: 0.2089

Epoch: 16


100%|██████████| 133/133 [00:00<00:00, 429.90it/s]



Train set: Average loss: 0.1920
Test set: Average loss: 0.2030

Epoch: 17


100%|██████████| 133/133 [00:00<00:00, 456.47it/s]



Train set: Average loss: 0.1856
Test set: Average loss: 0.1980

Epoch: 18


100%|██████████| 133/133 [00:00<00:00, 422.31it/s]



Train set: Average loss: 0.1800
Test set: Average loss: 0.1937

Epoch: 19


100%|██████████| 133/133 [00:00<00:00, 426.87it/s]



Train set: Average loss: 0.1753
Test set: Average loss: 0.1901

Epoch: 20


100%|██████████| 133/133 [00:00<00:00, 471.88it/s]



Train set: Average loss: 0.1711
Test set: Average loss: 0.1869

Epoch: 21


100%|██████████| 133/133 [00:00<00:00, 472.78it/s]



Train set: Average loss: 0.1675
Test set: Average loss: 0.1842

Epoch: 22


100%|██████████| 133/133 [00:00<00:00, 454.13it/s]



Train set: Average loss: 0.1643
Test set: Average loss: 0.1819

Epoch: 23


100%|██████████| 133/133 [00:00<00:00, 466.53it/s]



Train set: Average loss: 0.1615
Test set: Average loss: 0.1799

Epoch: 24


100%|██████████| 133/133 [00:00<00:00, 479.91it/s]



Train set: Average loss: 0.1591
Test set: Average loss: 0.1782

Epoch: 25


100%|██████████| 133/133 [00:00<00:00, 423.84it/s]



Train set: Average loss: 0.1570
Test set: Average loss: 0.1768

Epoch: 26


100%|██████████| 133/133 [00:00<00:00, 463.95it/s]



Train set: Average loss: 0.1551
Test set: Average loss: 0.1755

Epoch: 27


100%|██████████| 133/133 [00:00<00:00, 485.44it/s]



Train set: Average loss: 0.1534
Test set: Average loss: 0.1744

Epoch: 28


100%|██████████| 133/133 [00:00<00:00, 452.05it/s]



Train set: Average loss: 0.1520
Test set: Average loss: 0.1735

Epoch: 29


100%|██████████| 133/133 [00:00<00:00, 479.55it/s]



Train set: Average loss: 0.1507
Test set: Average loss: 0.1727

Epoch: 30


100%|██████████| 133/133 [00:00<00:00, 488.89it/s]



Train set: Average loss: 0.1496
Test set: Average loss: 0.1720

Epoch: 31


100%|██████████| 133/133 [00:00<00:00, 444.52it/s]



Train set: Average loss: 0.1486
Test set: Average loss: 0.1715

Epoch: 32


100%|██████████| 133/133 [00:00<00:00, 491.26it/s]



Train set: Average loss: 0.1477
Test set: Average loss: 0.1710

Epoch: 33


100%|██████████| 133/133 [00:00<00:00, 483.17it/s]



Train set: Average loss: 0.1469
Test set: Average loss: 0.1706

Epoch: 34


100%|██████████| 133/133 [00:00<00:00, 437.38it/s]



Train set: Average loss: 0.1462
Test set: Average loss: 0.1703

Epoch: 35


100%|██████████| 133/133 [00:00<00:00, 479.39it/s]



Train set: Average loss: 0.1456
Test set: Average loss: 0.1700

Epoch: 36


100%|██████████| 133/133 [00:00<00:00, 487.40it/s]



Train set: Average loss: 0.1451
Test set: Average loss: 0.1698

Epoch: 37


100%|██████████| 133/133 [00:00<00:00, 453.09it/s]



Train set: Average loss: 0.1446
Test set: Average loss: 0.1697

Epoch: 38


100%|██████████| 133/133 [00:00<00:00, 491.62it/s]



Train set: Average loss: 0.1442
Test set: Average loss: 0.1695

Epoch: 39


100%|██████████| 133/133 [00:00<00:00, 481.67it/s]



Train set: Average loss: 0.1438
Test set: Average loss: 0.1695

Epoch: 40


100%|██████████| 133/133 [00:00<00:00, 444.21it/s]



Train set: Average loss: 0.1435
Test set: Average loss: 0.1694

Epoch: 41


100%|██████████| 133/133 [00:00<00:00, 479.11it/s]



Train set: Average loss: 0.1433
Test set: Average loss: 0.1694

Epoch: 42


100%|██████████| 133/133 [00:00<00:00, 299.04it/s]



Train set: Average loss: 0.1428
Test set: Average loss: 0.1681

Epoch: 43


100%|██████████| 133/133 [00:00<00:00, 326.64it/s]



Train set: Average loss: 0.1394
Test set: Average loss: 0.1685

Epoch: 44


100%|██████████| 133/133 [00:00<00:00, 332.85it/s]



Train set: Average loss: 0.1342
Test set: Average loss: 0.1687

Epoch: 45


100%|██████████| 133/133 [00:00<00:00, 313.29it/s]



Train set: Average loss: 0.1329
Test set: Average loss: 0.1706

Epoch: 46


100%|██████████| 133/133 [00:00<00:00, 297.72it/s]



Train set: Average loss: 0.1308
Test set: Average loss: 0.1697

Epoch: 47


100%|██████████| 133/133 [00:00<00:00, 256.07it/s]



Train set: Average loss: 0.1296
Test set: Average loss: 0.1853

Epoch: 48


100%|██████████| 133/133 [00:00<00:00, 429.69it/s]



Train set: Average loss: 0.1291
Test set: Average loss: 0.1744

Epoch: 49


100%|██████████| 133/133 [00:00<00:00, 390.60it/s]



Train set: Average loss: 0.1284
Test set: Average loss: 0.1810

Epoch: 50


100%|██████████| 133/133 [00:00<00:00, 149.09it/s]



Train set: Average loss: 0.1275
Test set: Average loss: 0.1738

Epoch: 51


100%|██████████| 133/133 [00:00<00:00, 458.54it/s]



Train set: Average loss: 0.1265
Test set: Average loss: 0.1768

Epoch: 52


100%|██████████| 133/133 [00:00<00:00, 187.47it/s]



Train set: Average loss: 0.1255
Test set: Average loss: 0.1786

Epoch: 53


100%|██████████| 133/133 [00:00<00:00, 158.09it/s]



Train set: Average loss: 0.1246
Test set: Average loss: 0.1834

Epoch: 54


100%|██████████| 133/133 [00:00<00:00, 455.26it/s]



Train set: Average loss: 0.1240
Test set: Average loss: 0.1769

Epoch: 55


100%|██████████| 133/133 [00:00<00:00, 479.79it/s]



Train set: Average loss: 0.1237
Test set: Average loss: 0.1826

Epoch: 56


100%|██████████| 133/133 [00:00<00:00, 438.36it/s]



Train set: Average loss: 0.1225
Test set: Average loss: 0.1898

Epoch: 57


100%|██████████| 133/133 [00:00<00:00, 480.77it/s]



Train set: Average loss: 0.1218
Test set: Average loss: 0.1846

Epoch: 58


100%|██████████| 133/133 [00:00<00:00, 484.20it/s]



Train set: Average loss: 0.1216
Test set: Average loss: 0.1820

Epoch: 59


100%|██████████| 133/133 [00:00<00:00, 432.23it/s]



Train set: Average loss: 0.1207
Test set: Average loss: 0.1940

Epoch: 60


100%|██████████| 133/133 [00:00<00:00, 467.96it/s]



Train set: Average loss: 0.1194
Test set: Average loss: 0.1890
Training is ended!

BEST EPOCH:  42 with Loss:  0.1681321774548185


Реализована базовая модель полносвязной нейронной сети

Проверим гиперпараметры, выберем лучшие варианты

1. Количество нейронов на 1 слое

варианты:

X - количество параметров - 19шт
Нужна убывающая регрессия (для уменьшения количества нейронов на каждом последующем слое), идем по степеням двойки

X*1 - 19, берем 32 нейрона в базовом слое [32, 16, 8]

X*2 - 38, берем 64 нейрона в базовом слое [64, 32, 16]

X*4 - 76, берем 128 нейронов в базовом слое [128, 64, 32]

X*7 - 133, берем 258 нейронов в базовом слое [256, 128, 64]

X*8 - 152, дальше нет смысла идти, высокий шанс переобучения

In [82]:
class CFG:

# Задаем параметры нашего эксперимента
  feature_cols = ['engine_capacity', 'road_quality_moderate', 'slope_flat',
                'motorway', 'rural', 'more_than_one_lane', 'congested',
                'speed_limit_mean', 'weather_temperature_mean', 'total_distance',
                'sum_roundabout', 'sum_traffic_signal', 'sum_stop_sign',
                'sum_yield_sign', 'sum_pedestrian_crossing_sign',
                'sum_animal_crossing_sign', 'speeding_serious',
                'harsh_acceleration', 'harsh_braking']

  input_dim = len(feature_cols)
  hidden_dims =  [128, 64, 32]
  dropout_rate = 0.2
  learning_rate = 0.001
  num_epochs = 60
class Regression(nn.Module): # наследуемся от класса nn.Module
    def __init__(self):
        super(Regression,self).__init__()
        # организуем 3 скрытых слоя
        hidden_1 =  CFG.hidden_dims[0]
        hidden_2 = CFG.hidden_dims[1]
        hidden_3 = CFG.hidden_dims[2]
        #
        input_dim = 19
        self.net = torch.nn.Sequential(
                      torch.nn.Linear(input_dim, hidden_1),
                      torch.nn.ReLU(),
                      torch.nn.Linear(hidden_1, hidden_2),
                      torch.nn.ReLU(),
                      torch.nn.Linear(hidden_2, hidden_3),
                      torch.nn.ReLU(),
                      torch.nn.Linear(hidden_3, 1),
                    )

    def forward(self,x):
        x = torch.exp(self.net(x))
        return x
model = Regression()
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = model.to(device) # переводим модель на GPU, не получилось, поэтому CPU
print(model) # посмотрим на нашу модель

# функция потерь
criterion = nn.MSELoss()

optimizer = torch.optim.Adam(model.parameters(),lr = 0.001)

# функция обучения модели
def train(model, device, train_loader, optimizer, criterion, epoch):
    model.train() # обязательно переводим в режим обучения
    train_loss_sum = 0


    n_ex = len(train_loader)

    for batch_idx, (data, target) in tqdm(enumerate(train_loader), total=n_ex):
        data, target = data.to(device), target.to(device) # переводим картинки и таргеты на GPU
        # обнуляем градиенты!
        optimizer.zero_grad()
        # прямой проход
        output = model(data)

        train_loss = criterion(output, target) # считаем значение функции потерь
        # обратный проход
        train_loss.backward()
        # делаем градиентный шаг оптимизатором
        optimizer.step()
        # считаем метрики и лосс
        train_loss_sum += train_loss.item() * data.size(0)

    tqdm.write('\nTrain set: Average loss: {:.4f}'.format(
        train_loss_sum / len(train_loader.dataset)))
# функция тестирования
def test(model, device, test_loader, criterion):
    model.eval() # переводем модель в режим инференса
    test_loss_sum = 0

    all_predictions = []
    all_targets = []

    # показываем, что обученич нет и градиенты не обновляются
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss = criterion(output, target) # считаем значение функции потерь
            test_loss_sum += test_loss.item() * data.size(0)

            predictions = output.cpu().numpy().flatten()
            targets = target.cpu().numpy().flatten()

            all_predictions.extend(predictions.tolist())
            all_targets.extend(targets.tolist())


            # считаем метрики
    tqdm.write('Test set: Average loss: {:.4f}'.format(
       test_loss_sum / len(test_loader.dataset)))
    return test_loss_sum / len(test_loader.dataset)

# основная функция для экспериментов
def main(model):

    seed_everything(30)

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu") # выделили устройство
    model = model.to(device)

    losses = []
    for epoch in range(1, CFG.num_epochs + 1): # цикл на эпохи
        print('\nEpoch:', epoch)
        train(model, device, train_loader, optimizer, criterion, epoch)
        losses.append(test(model, device, test_loader, criterion))
    print('Training is ended!')
    best_epoch = np.argmin(losses)+1
    print()
    print('BEST EPOCH: ', best_epoch, 'with Loss: ',min(losses))



Regression(
  (net): Sequential(
    (0): Linear(in_features=19, out_features=128, bias=True)
    (1): ReLU()
    (2): Linear(in_features=128, out_features=64, bias=True)
    (3): ReLU()
    (4): Linear(in_features=64, out_features=32, bias=True)
    (5): ReLU()
    (6): Linear(in_features=32, out_features=1, bias=True)
  )
)


In [83]:
main(model)


Epoch: 1


100%|██████████| 133/133 [00:00<00:00, 370.16it/s]



Train set: Average loss: 0.2179
Test set: Average loss: 0.1726

Epoch: 2


100%|██████████| 133/133 [00:00<00:00, 278.40it/s]



Train set: Average loss: 0.1404
Test set: Average loss: 0.1698

Epoch: 3


100%|██████████| 133/133 [00:00<00:00, 244.22it/s]



Train set: Average loss: 0.1331
Test set: Average loss: 0.1698

Epoch: 4


100%|██████████| 133/133 [00:00<00:00, 207.59it/s]



Train set: Average loss: 0.1279
Test set: Average loss: 0.1705

Epoch: 5


100%|██████████| 133/133 [00:00<00:00, 164.58it/s]



Train set: Average loss: 0.1239
Test set: Average loss: 0.1742

Epoch: 6


100%|██████████| 133/133 [00:01<00:00, 95.23it/s]



Train set: Average loss: 0.1193
Test set: Average loss: 0.1736

Epoch: 7


100%|██████████| 133/133 [00:01<00:00, 78.19it/s]



Train set: Average loss: 0.1173
Test set: Average loss: 0.1750

Epoch: 8


100%|██████████| 133/133 [00:00<00:00, 184.46it/s]



Train set: Average loss: 0.1141
Test set: Average loss: 0.1789

Epoch: 9


100%|██████████| 133/133 [00:00<00:00, 188.20it/s]



Train set: Average loss: 0.1105
Test set: Average loss: 0.1802

Epoch: 10


100%|██████████| 133/133 [00:00<00:00, 185.31it/s]



Train set: Average loss: 0.1071
Test set: Average loss: 0.1832

Epoch: 11


100%|██████████| 133/133 [00:01<00:00, 127.90it/s]



Train set: Average loss: 0.1051
Test set: Average loss: 0.1821

Epoch: 12


100%|██████████| 133/133 [00:00<00:00, 176.12it/s]



Train set: Average loss: 0.1018
Test set: Average loss: 0.1875

Epoch: 13


100%|██████████| 133/133 [00:00<00:00, 181.76it/s]



Train set: Average loss: 0.0996
Test set: Average loss: 0.1859

Epoch: 14


100%|██████████| 133/133 [00:00<00:00, 139.01it/s]



Train set: Average loss: 0.0973
Test set: Average loss: 0.1856

Epoch: 15


100%|██████████| 133/133 [00:00<00:00, 168.78it/s]



Train set: Average loss: 0.0947
Test set: Average loss: 0.1884

Epoch: 16


100%|██████████| 133/133 [00:00<00:00, 141.29it/s]



Train set: Average loss: 0.0916
Test set: Average loss: 0.2040

Epoch: 17


100%|██████████| 133/133 [00:00<00:00, 169.07it/s]



Train set: Average loss: 0.0889
Test set: Average loss: 0.2014

Epoch: 18


100%|██████████| 133/133 [00:01<00:00, 86.65it/s]



Train set: Average loss: 0.0880
Test set: Average loss: 0.1949

Epoch: 19


100%|██████████| 133/133 [00:01<00:00, 86.64it/s]



Train set: Average loss: 0.0839
Test set: Average loss: 0.2020

Epoch: 20


100%|██████████| 133/133 [00:00<00:00, 137.28it/s]



Train set: Average loss: 0.0831
Test set: Average loss: 0.2100

Epoch: 21


100%|██████████| 133/133 [00:00<00:00, 157.29it/s]



Train set: Average loss: 0.0809
Test set: Average loss: 0.2097

Epoch: 22


100%|██████████| 133/133 [00:00<00:00, 198.35it/s]



Train set: Average loss: 0.0767
Test set: Average loss: 0.1973

Epoch: 23


100%|██████████| 133/133 [00:00<00:00, 178.34it/s]



Train set: Average loss: 0.0745
Test set: Average loss: 0.2180

Epoch: 24


100%|██████████| 133/133 [00:00<00:00, 218.20it/s]



Train set: Average loss: 0.0715
Test set: Average loss: 0.2085

Epoch: 25


100%|██████████| 133/133 [00:00<00:00, 139.97it/s]



Train set: Average loss: 0.0697
Test set: Average loss: 0.2113

Epoch: 26


100%|██████████| 133/133 [00:00<00:00, 182.98it/s]



Train set: Average loss: 0.0718
Test set: Average loss: 0.2221

Epoch: 27


100%|██████████| 133/133 [00:00<00:00, 212.95it/s]



Train set: Average loss: 0.0672
Test set: Average loss: 0.2088

Epoch: 28


100%|██████████| 133/133 [00:00<00:00, 286.10it/s]



Train set: Average loss: 0.0666
Test set: Average loss: 0.2134

Epoch: 29


100%|██████████| 133/133 [00:00<00:00, 386.69it/s]



Train set: Average loss: 0.0634
Test set: Average loss: 0.2193

Epoch: 30


100%|██████████| 133/133 [00:00<00:00, 406.45it/s]



Train set: Average loss: 0.0612
Test set: Average loss: 0.2169

Epoch: 31


100%|██████████| 133/133 [00:00<00:00, 415.75it/s]



Train set: Average loss: 0.0585
Test set: Average loss: 0.2214

Epoch: 32


100%|██████████| 133/133 [00:00<00:00, 426.42it/s]



Train set: Average loss: 0.0607
Test set: Average loss: 0.2094

Epoch: 33


100%|██████████| 133/133 [00:00<00:00, 437.14it/s]



Train set: Average loss: 0.0574
Test set: Average loss: 0.2097

Epoch: 34


100%|██████████| 133/133 [00:00<00:00, 398.83it/s]



Train set: Average loss: 0.0569
Test set: Average loss: 0.2165

Epoch: 35


100%|██████████| 133/133 [00:00<00:00, 263.13it/s]



Train set: Average loss: 0.0541
Test set: Average loss: 0.2321

Epoch: 36


100%|██████████| 133/133 [00:00<00:00, 295.29it/s]



Train set: Average loss: 0.0537
Test set: Average loss: 0.2132

Epoch: 37


100%|██████████| 133/133 [00:00<00:00, 311.85it/s]



Train set: Average loss: 0.0538
Test set: Average loss: 0.2189

Epoch: 38


100%|██████████| 133/133 [00:00<00:00, 271.38it/s]



Train set: Average loss: 0.0513
Test set: Average loss: 0.2470

Epoch: 39


100%|██████████| 133/133 [00:00<00:00, 259.23it/s]



Train set: Average loss: 0.0493
Test set: Average loss: 0.2250

Epoch: 40


100%|██████████| 133/133 [00:00<00:00, 290.11it/s]



Train set: Average loss: 0.0483
Test set: Average loss: 0.2407

Epoch: 41


100%|██████████| 133/133 [00:00<00:00, 420.00it/s]



Train set: Average loss: 0.0506
Test set: Average loss: 0.2300

Epoch: 42


100%|██████████| 133/133 [00:00<00:00, 415.05it/s]



Train set: Average loss: 0.0487
Test set: Average loss: 0.2340

Epoch: 43


100%|██████████| 133/133 [00:00<00:00, 399.81it/s]



Train set: Average loss: 0.0477
Test set: Average loss: 0.2318

Epoch: 44


100%|██████████| 133/133 [00:00<00:00, 405.16it/s]



Train set: Average loss: 0.0451
Test set: Average loss: 0.2239

Epoch: 45


100%|██████████| 133/133 [00:00<00:00, 399.70it/s]



Train set: Average loss: 0.0446
Test set: Average loss: 0.2389

Epoch: 46


100%|██████████| 133/133 [00:00<00:00, 411.94it/s]



Train set: Average loss: 0.0428
Test set: Average loss: 0.2365

Epoch: 47


100%|██████████| 133/133 [00:00<00:00, 383.82it/s]



Train set: Average loss: 0.0416
Test set: Average loss: 0.2324

Epoch: 48


100%|██████████| 133/133 [00:00<00:00, 398.46it/s]



Train set: Average loss: 0.0419
Test set: Average loss: 0.2394

Epoch: 49


100%|██████████| 133/133 [00:00<00:00, 403.19it/s]



Train set: Average loss: 0.0410
Test set: Average loss: 0.2326

Epoch: 50


100%|██████████| 133/133 [00:00<00:00, 405.01it/s]



Train set: Average loss: 0.0405
Test set: Average loss: 0.2404

Epoch: 51


100%|██████████| 133/133 [00:00<00:00, 423.45it/s]



Train set: Average loss: 0.0414
Test set: Average loss: 0.2371

Epoch: 52


100%|██████████| 133/133 [00:00<00:00, 345.95it/s]



Train set: Average loss: 0.0391
Test set: Average loss: 0.2353

Epoch: 53


100%|██████████| 133/133 [00:00<00:00, 359.81it/s]



Train set: Average loss: 0.0392
Test set: Average loss: 0.2569

Epoch: 54


100%|██████████| 133/133 [00:00<00:00, 402.17it/s]



Train set: Average loss: 0.0396
Test set: Average loss: 0.2200

Epoch: 55


100%|██████████| 133/133 [00:00<00:00, 405.82it/s]



Train set: Average loss: 0.0384
Test set: Average loss: 0.2430

Epoch: 56


100%|██████████| 133/133 [00:00<00:00, 418.45it/s]



Train set: Average loss: 0.0353
Test set: Average loss: 0.2559

Epoch: 57


100%|██████████| 133/133 [00:00<00:00, 410.85it/s]



Train set: Average loss: 0.0374
Test set: Average loss: 0.2352

Epoch: 58


100%|██████████| 133/133 [00:00<00:00, 394.04it/s]



Train set: Average loss: 0.0354
Test set: Average loss: 0.2447

Epoch: 59


100%|██████████| 133/133 [00:00<00:00, 408.61it/s]



Train set: Average loss: 0.0327
Test set: Average loss: 0.2385

Epoch: 60


100%|██████████| 133/133 [00:00<00:00, 412.27it/s]


Train set: Average loss: 0.0357
Test set: Average loss: 0.2629
Training is ended!

BEST EPOCH:  2 with Loss:  0.16978198520100207


In [84]:
class CFG:

# Задаем параметры нашего эксперимента
  feature_cols = ['engine_capacity', 'road_quality_moderate', 'slope_flat',
                'motorway', 'rural', 'more_than_one_lane', 'congested',
                'speed_limit_mean', 'weather_temperature_mean', 'total_distance',
                'sum_roundabout', 'sum_traffic_signal', 'sum_stop_sign',
                'sum_yield_sign', 'sum_pedestrian_crossing_sign',
                'sum_animal_crossing_sign', 'speeding_serious',
                'harsh_acceleration', 'harsh_braking']

  input_dim = len(feature_cols)
  hidden_dims =   [64, 32, 16]
  dropout_rate = 0.2
  learning_rate = 0.001
  num_epochs = 60
class Regression(nn.Module): # наследуемся от класса nn.Module
    def __init__(self):
        super(Regression,self).__init__()
        # организуем 3 скрытых слоя
        hidden_1 =  CFG.hidden_dims[0]
        hidden_2 = CFG.hidden_dims[1]
        hidden_3 = CFG.hidden_dims[2]
        #
        input_dim = 19
        self.net = torch.nn.Sequential(
                      torch.nn.Linear(input_dim, hidden_1),
                      torch.nn.ReLU(),
                      torch.nn.Linear(hidden_1, hidden_2),
                      torch.nn.ReLU(),
                      torch.nn.Linear(hidden_2, hidden_3),
                      torch.nn.ReLU(),
                      torch.nn.Linear(hidden_3, 1),
                    )

    def forward(self,x):
        x = torch.exp(self.net(x))
        return x
model = Regression()
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = model.to(device) # переводим модель на GPU, не получилось, поэтому CPU
print(model) # посмотрим на нашу модель

# функция потерь
criterion = nn.MSELoss()

optimizer = torch.optim.Adam(model.parameters(),lr = 0.001)

# функция обучения модели
def train(model, device, train_loader, optimizer, criterion, epoch):
    model.train() # обязательно переводим в режим обучения
    train_loss_sum = 0


    n_ex = len(train_loader)

    for batch_idx, (data, target) in tqdm(enumerate(train_loader), total=n_ex):
        data, target = data.to(device), target.to(device) # переводим картинки и таргеты на GPU
        # обнуляем градиенты!
        optimizer.zero_grad()
        # прямой проход
        output = model(data)

        train_loss = criterion(output, target) # считаем значение функции потерь
        # обратный проход
        train_loss.backward()
        # делаем градиентный шаг оптимизатором
        optimizer.step()
        # считаем метрики и лосс
        train_loss_sum += train_loss.item() * data.size(0)

    tqdm.write('\nTrain set: Average loss: {:.4f}'.format(
        train_loss_sum / len(train_loader.dataset)))
# функция тестирования
def test(model, device, test_loader, criterion):
    model.eval() # переводем модель в режим инференса
    test_loss_sum = 0

    all_predictions = []
    all_targets = []

    # показываем, что обученич нет и градиенты не обновляются
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss = criterion(output, target) # считаем значение функции потерь
            test_loss_sum += test_loss.item() * data.size(0)

            predictions = output.cpu().numpy().flatten()
            targets = target.cpu().numpy().flatten()

            all_predictions.extend(predictions.tolist())
            all_targets.extend(targets.tolist())


            # считаем метрики
    tqdm.write('Test set: Average loss: {:.4f}'.format(
       test_loss_sum / len(test_loader.dataset)))
    return test_loss_sum / len(test_loader.dataset)

# основная функция для экспериментов
def main(model):

    seed_everything(30)

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu") # выделили устройство
    model = model.to(device)

    losses = []
    for epoch in range(1, CFG.num_epochs + 1): # цикл на эпохи
        print('\nEpoch:', epoch)
        train(model, device, train_loader, optimizer, criterion, epoch)
        losses.append(test(model, device, test_loader, criterion))
    print('Training is ended!')
    best_epoch = np.argmin(losses)+1
    print()
    print('BEST EPOCH: ', best_epoch, 'with Loss: ',min(losses))



Regression(
  (net): Sequential(
    (0): Linear(in_features=19, out_features=64, bias=True)
    (1): ReLU()
    (2): Linear(in_features=64, out_features=32, bias=True)
    (3): ReLU()
    (4): Linear(in_features=32, out_features=16, bias=True)
    (5): ReLU()
    (6): Linear(in_features=16, out_features=1, bias=True)
  )
)


In [85]:
main(model)


Epoch: 1


100%|██████████| 133/133 [00:00<00:00, 487.98it/s]



Train set: Average loss: 0.2092
Test set: Average loss: 0.1726

Epoch: 2


100%|██████████| 133/133 [00:00<00:00, 476.04it/s]



Train set: Average loss: 0.1400
Test set: Average loss: 0.1705

Epoch: 3


100%|██████████| 133/133 [00:00<00:00, 442.97it/s]



Train set: Average loss: 0.1339
Test set: Average loss: 0.1703

Epoch: 4


100%|██████████| 133/133 [00:00<00:00, 450.79it/s]



Train set: Average loss: 0.1295
Test set: Average loss: 0.1705

Epoch: 5


100%|██████████| 133/133 [00:00<00:00, 467.85it/s]



Train set: Average loss: 0.1267
Test set: Average loss: 0.1705

Epoch: 6


100%|██████████| 133/133 [00:00<00:00, 454.74it/s]



Train set: Average loss: 0.1241
Test set: Average loss: 0.1715

Epoch: 7


100%|██████████| 133/133 [00:00<00:00, 486.68it/s]



Train set: Average loss: 0.1217
Test set: Average loss: 0.1717

Epoch: 8


100%|██████████| 133/133 [00:00<00:00, 471.78it/s]



Train set: Average loss: 0.1200
Test set: Average loss: 0.1768

Epoch: 9


100%|██████████| 133/133 [00:00<00:00, 445.16it/s]



Train set: Average loss: 0.1173
Test set: Average loss: 0.1758

Epoch: 10


100%|██████████| 133/133 [00:00<00:00, 472.38it/s]



Train set: Average loss: 0.1155
Test set: Average loss: 0.1777

Epoch: 11


100%|██████████| 133/133 [00:00<00:00, 456.54it/s]



Train set: Average loss: 0.1134
Test set: Average loss: 0.1797

Epoch: 12


100%|██████████| 133/133 [00:00<00:00, 316.30it/s]



Train set: Average loss: 0.1105
Test set: Average loss: 0.1848

Epoch: 13


100%|██████████| 133/133 [00:00<00:00, 332.83it/s]



Train set: Average loss: 0.1088
Test set: Average loss: 0.1870

Epoch: 14


100%|██████████| 133/133 [00:00<00:00, 308.23it/s]



Train set: Average loss: 0.1065
Test set: Average loss: 0.1831

Epoch: 15


100%|██████████| 133/133 [00:00<00:00, 326.60it/s]



Train set: Average loss: 0.1059
Test set: Average loss: 0.1853

Epoch: 16


100%|██████████| 133/133 [00:00<00:00, 291.86it/s]



Train set: Average loss: 0.1036
Test set: Average loss: 0.2007

Epoch: 17


100%|██████████| 133/133 [00:00<00:00, 280.71it/s]



Train set: Average loss: 0.1024
Test set: Average loss: 0.1973

Epoch: 18


100%|██████████| 133/133 [00:00<00:00, 430.11it/s]



Train set: Average loss: 0.1013
Test set: Average loss: 0.1953

Epoch: 19


100%|██████████| 133/133 [00:00<00:00, 451.87it/s]



Train set: Average loss: 0.0995
Test set: Average loss: 0.1958

Epoch: 20


100%|██████████| 133/133 [00:00<00:00, 459.66it/s]



Train set: Average loss: 0.0976
Test set: Average loss: 0.1999

Epoch: 21


100%|██████████| 133/133 [00:00<00:00, 446.21it/s]



Train set: Average loss: 0.0964
Test set: Average loss: 0.2106

Epoch: 22


100%|██████████| 133/133 [00:00<00:00, 455.36it/s]



Train set: Average loss: 0.0953
Test set: Average loss: 0.1950

Epoch: 23


100%|██████████| 133/133 [00:00<00:00, 444.94it/s]



Train set: Average loss: 0.0952
Test set: Average loss: 0.1960

Epoch: 24


100%|██████████| 133/133 [00:00<00:00, 473.51it/s]



Train set: Average loss: 0.0926
Test set: Average loss: 0.1907

Epoch: 25


100%|██████████| 133/133 [00:00<00:00, 417.11it/s]



Train set: Average loss: 0.0926
Test set: Average loss: 0.2005

Epoch: 26


100%|██████████| 133/133 [00:00<00:00, 202.32it/s]



Train set: Average loss: 0.0902
Test set: Average loss: 0.1945

Epoch: 27


100%|██████████| 133/133 [00:00<00:00, 192.59it/s]



Train set: Average loss: 0.0893
Test set: Average loss: 0.2022

Epoch: 28


100%|██████████| 133/133 [00:00<00:00, 221.16it/s]



Train set: Average loss: 0.0889
Test set: Average loss: 0.2049

Epoch: 29


100%|██████████| 133/133 [00:00<00:00, 173.95it/s]



Train set: Average loss: 0.0867
Test set: Average loss: 0.2017

Epoch: 30


100%|██████████| 133/133 [00:00<00:00, 165.28it/s]



Train set: Average loss: 0.0857
Test set: Average loss: 0.2071

Epoch: 31


100%|██████████| 133/133 [00:00<00:00, 196.66it/s]



Train set: Average loss: 0.0840
Test set: Average loss: 0.1976

Epoch: 32


100%|██████████| 133/133 [00:00<00:00, 208.42it/s]



Train set: Average loss: 0.0832
Test set: Average loss: 0.1923

Epoch: 33


100%|██████████| 133/133 [00:00<00:00, 223.34it/s]



Train set: Average loss: 0.0817
Test set: Average loss: 0.1967

Epoch: 34


100%|██████████| 133/133 [00:00<00:00, 166.49it/s]



Train set: Average loss: 0.0813
Test set: Average loss: 0.2070

Epoch: 35


100%|██████████| 133/133 [00:02<00:00, 66.34it/s]



Train set: Average loss: 0.0798
Test set: Average loss: 0.2042

Epoch: 36


100%|██████████| 133/133 [00:00<00:00, 157.26it/s]



Train set: Average loss: 0.0778
Test set: Average loss: 0.2055

Epoch: 37


100%|██████████| 133/133 [00:00<00:00, 193.73it/s]



Train set: Average loss: 0.0769
Test set: Average loss: 0.1986

Epoch: 38


100%|██████████| 133/133 [00:00<00:00, 390.17it/s]



Train set: Average loss: 0.0780
Test set: Average loss: 0.2086

Epoch: 39


100%|██████████| 133/133 [00:00<00:00, 393.98it/s]



Train set: Average loss: 0.0755
Test set: Average loss: 0.2067

Epoch: 40


100%|██████████| 133/133 [00:00<00:00, 442.30it/s]



Train set: Average loss: 0.0743
Test set: Average loss: 0.2143

Epoch: 41


100%|██████████| 133/133 [00:00<00:00, 441.62it/s]



Train set: Average loss: 0.0736
Test set: Average loss: 0.2071

Epoch: 42


100%|██████████| 133/133 [00:00<00:00, 447.18it/s]



Train set: Average loss: 0.0718
Test set: Average loss: 0.2011

Epoch: 43


100%|██████████| 133/133 [00:00<00:00, 465.06it/s]



Train set: Average loss: 0.0716
Test set: Average loss: 0.2112

Epoch: 44


100%|██████████| 133/133 [00:00<00:00, 446.59it/s]



Train set: Average loss: 0.0709
Test set: Average loss: 0.2103

Epoch: 45


100%|██████████| 133/133 [00:00<00:00, 459.89it/s]



Train set: Average loss: 0.0699
Test set: Average loss: 0.2213

Epoch: 46


100%|██████████| 133/133 [00:00<00:00, 474.92it/s]



Train set: Average loss: 0.0700
Test set: Average loss: 0.2159

Epoch: 47


100%|██████████| 133/133 [00:00<00:00, 429.47it/s]



Train set: Average loss: 0.0687
Test set: Average loss: 0.2195

Epoch: 48


100%|██████████| 133/133 [00:00<00:00, 462.95it/s]



Train set: Average loss: 0.0696
Test set: Average loss: 0.2057

Epoch: 49


100%|██████████| 133/133 [00:00<00:00, 450.00it/s]



Train set: Average loss: 0.0679
Test set: Average loss: 0.2218

Epoch: 50


100%|██████████| 133/133 [00:00<00:00, 426.57it/s]



Train set: Average loss: 0.0673
Test set: Average loss: 0.2194

Epoch: 51


100%|██████████| 133/133 [00:00<00:00, 444.74it/s]



Train set: Average loss: 0.0657
Test set: Average loss: 0.2031

Epoch: 52


100%|██████████| 133/133 [00:00<00:00, 457.04it/s]



Train set: Average loss: 0.0672
Test set: Average loss: 0.2086

Epoch: 53


100%|██████████| 133/133 [00:00<00:00, 412.99it/s]



Train set: Average loss: 0.0655
Test set: Average loss: 0.2229

Epoch: 54


100%|██████████| 133/133 [00:00<00:00, 443.82it/s]



Train set: Average loss: 0.0653
Test set: Average loss: 0.2088

Epoch: 55


100%|██████████| 133/133 [00:00<00:00, 452.08it/s]



Train set: Average loss: 0.0643
Test set: Average loss: 0.2210

Epoch: 56


100%|██████████| 133/133 [00:00<00:00, 428.30it/s]



Train set: Average loss: 0.0632
Test set: Average loss: 0.2088

Epoch: 57


100%|██████████| 133/133 [00:00<00:00, 447.88it/s]



Train set: Average loss: 0.0618
Test set: Average loss: 0.2107

Epoch: 58


100%|██████████| 133/133 [00:00<00:00, 433.49it/s]



Train set: Average loss: 0.0616
Test set: Average loss: 0.2265

Epoch: 59


100%|██████████| 133/133 [00:00<00:00, 403.67it/s]



Train set: Average loss: 0.0600
Test set: Average loss: 0.2343

Epoch: 60


100%|██████████| 133/133 [00:00<00:00, 450.30it/s]



Train set: Average loss: 0.0611
Test set: Average loss: 0.2100
Training is ended!

BEST EPOCH:  3 with Loss:  0.17034084483231124


In [86]:
class CFG:

# Задаем параметры нашего эксперимента
  feature_cols = ['engine_capacity', 'road_quality_moderate', 'slope_flat',
                'motorway', 'rural', 'more_than_one_lane', 'congested',
                'speed_limit_mean', 'weather_temperature_mean', 'total_distance',
                'sum_roundabout', 'sum_traffic_signal', 'sum_stop_sign',
                'sum_yield_sign', 'sum_pedestrian_crossing_sign',
                'sum_animal_crossing_sign', 'speeding_serious',
                'harsh_acceleration', 'harsh_braking']

  input_dim = len(feature_cols)
  hidden_dims =   [256, 128, 64]
  dropout_rate = 0.2
  learning_rate = 0.001
  num_epochs = 60
class Regression(nn.Module): # наследуемся от класса nn.Module
    def __init__(self):
        super(Regression,self).__init__()
        # организуем 3 скрытых слоя
        hidden_1 =  CFG.hidden_dims[0]
        hidden_2 = CFG.hidden_dims[1]
        hidden_3 = CFG.hidden_dims[2]
        #
        input_dim = 19
        self.net = torch.nn.Sequential(
                      torch.nn.Linear(input_dim, hidden_1),
                      torch.nn.ReLU(),
                      torch.nn.Linear(hidden_1, hidden_2),
                      torch.nn.ReLU(),
                      torch.nn.Linear(hidden_2, hidden_3),
                      torch.nn.ReLU(),
                      torch.nn.Linear(hidden_3, 1),
                    )

    def forward(self,x):
        x = torch.exp(self.net(x))
        return x
model = Regression()
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model = model.to(device) # переводим модель на GPU, не получилось, поэтому CPU
print(model) # посмотрим на нашу модель

# функция потерь
criterion = nn.MSELoss()

optimizer = torch.optim.Adam(model.parameters(),lr = 0.001)

# функция обучения модели
def train(model, device, train_loader, optimizer, criterion, epoch):
    model.train() # обязательно переводим в режим обучения
    train_loss_sum = 0


    n_ex = len(train_loader)

    for batch_idx, (data, target) in tqdm(enumerate(train_loader), total=n_ex):
        data, target = data.to(device), target.to(device) # переводим картинки и таргеты на GPU
        # обнуляем градиенты!
        optimizer.zero_grad()
        # прямой проход
        output = model(data)

        train_loss = criterion(output, target) # считаем значение функции потерь
        # обратный проход
        train_loss.backward()
        # делаем градиентный шаг оптимизатором
        optimizer.step()
        # считаем метрики и лосс
        train_loss_sum += train_loss.item() * data.size(0)

    tqdm.write('\nTrain set: Average loss: {:.4f}'.format(
        train_loss_sum / len(train_loader.dataset)))
# функция тестирования
def test(model, device, test_loader, criterion):
    model.eval() # переводем модель в режим инференса
    test_loss_sum = 0

    all_predictions = []
    all_targets = []

    # показываем, что обученич нет и градиенты не обновляются
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss = criterion(output, target) # считаем значение функции потерь
            test_loss_sum += test_loss.item() * data.size(0)

            predictions = output.cpu().numpy().flatten()
            targets = target.cpu().numpy().flatten()

            all_predictions.extend(predictions.tolist())
            all_targets.extend(targets.tolist())


            # считаем метрики
    tqdm.write('Test set: Average loss: {:.4f}'.format(
       test_loss_sum / len(test_loader.dataset)))
    return test_loss_sum / len(test_loader.dataset)

# основная функция для экспериментов
def main(model):

    seed_everything(30)

    device = torch.device("cuda" if torch.cuda.is_available() else "cpu") # выделили устройство
    model = model.to(device)

    losses = []
    for epoch in range(1, CFG.num_epochs + 1): # цикл на эпохи
        print('\nEpoch:', epoch)
        train(model, device, train_loader, optimizer, criterion, epoch)
        losses.append(test(model, device, test_loader, criterion))
    print('Training is ended!')
    best_epoch = np.argmin(losses)+1
    print()
    print('BEST EPOCH: ', best_epoch, 'with Loss: ',min(losses))



Regression(
  (net): Sequential(
    (0): Linear(in_features=19, out_features=256, bias=True)
    (1): ReLU()
    (2): Linear(in_features=256, out_features=128, bias=True)
    (3): ReLU()
    (4): Linear(in_features=128, out_features=64, bias=True)
    (5): ReLU()
    (6): Linear(in_features=64, out_features=1, bias=True)
  )
)


In [87]:
main(model)


Epoch: 1


100%|██████████| 133/133 [00:00<00:00, 340.01it/s]



Train set: Average loss: 0.1677
Test set: Average loss: 0.1710

Epoch: 2


100%|██████████| 133/133 [00:00<00:00, 335.25it/s]



Train set: Average loss: 0.1345
Test set: Average loss: 0.1714

Epoch: 3


100%|██████████| 133/133 [00:00<00:00, 341.29it/s]



Train set: Average loss: 0.1267
Test set: Average loss: 0.1769

Epoch: 4


100%|██████████| 133/133 [00:00<00:00, 329.43it/s]



Train set: Average loss: 0.1209
Test set: Average loss: 0.1777

Epoch: 5


100%|██████████| 133/133 [00:00<00:00, 333.98it/s]



Train set: Average loss: 0.1162
Test set: Average loss: 0.1813

Epoch: 6


100%|██████████| 133/133 [00:00<00:00, 205.12it/s]



Train set: Average loss: 0.1121
Test set: Average loss: 0.1814

Epoch: 7


100%|██████████| 133/133 [00:00<00:00, 205.80it/s]



Train set: Average loss: 0.1079
Test set: Average loss: 0.1861

Epoch: 8


100%|██████████| 133/133 [00:00<00:00, 215.48it/s]



Train set: Average loss: 0.1042
Test set: Average loss: 0.1878

Epoch: 9


100%|██████████| 133/133 [00:00<00:00, 194.60it/s]



Train set: Average loss: 0.1010
Test set: Average loss: 0.1988

Epoch: 10


100%|██████████| 133/133 [00:00<00:00, 300.80it/s]



Train set: Average loss: 0.0972
Test set: Average loss: 0.1990

Epoch: 11


100%|██████████| 133/133 [00:00<00:00, 296.40it/s]



Train set: Average loss: 0.0938
Test set: Average loss: 0.1957

Epoch: 12


100%|██████████| 133/133 [00:00<00:00, 277.30it/s]



Train set: Average loss: 0.0895
Test set: Average loss: 0.2005

Epoch: 13


100%|██████████| 133/133 [00:00<00:00, 261.98it/s]



Train set: Average loss: 0.0846
Test set: Average loss: 0.1945

Epoch: 14


100%|██████████| 133/133 [00:00<00:00, 301.73it/s]



Train set: Average loss: 0.0837
Test set: Average loss: 0.2047

Epoch: 15


100%|██████████| 133/133 [00:00<00:00, 301.77it/s]



Train set: Average loss: 0.0827
Test set: Average loss: 0.2020

Epoch: 16


100%|██████████| 133/133 [00:00<00:00, 306.42it/s]



Train set: Average loss: 0.0779
Test set: Average loss: 0.2103

Epoch: 17


100%|██████████| 133/133 [00:00<00:00, 305.49it/s]



Train set: Average loss: 0.0770
Test set: Average loss: 0.2135

Epoch: 18


100%|██████████| 133/133 [00:00<00:00, 298.86it/s]



Train set: Average loss: 0.0732
Test set: Average loss: 0.2123

Epoch: 19


100%|██████████| 133/133 [00:00<00:00, 159.02it/s]



Train set: Average loss: 0.0708
Test set: Average loss: 0.2256

Epoch: 20


100%|██████████| 133/133 [00:00<00:00, 174.83it/s]



Train set: Average loss: 0.0696
Test set: Average loss: 0.2427

Epoch: 21


100%|██████████| 133/133 [00:00<00:00, 181.62it/s]



Train set: Average loss: 0.0661
Test set: Average loss: 0.2484

Epoch: 22


100%|██████████| 133/133 [00:00<00:00, 167.17it/s]



Train set: Average loss: 0.0632
Test set: Average loss: 0.2226

Epoch: 23


100%|██████████| 133/133 [00:00<00:00, 161.10it/s]



Train set: Average loss: 0.0647
Test set: Average loss: 0.2312

Epoch: 24


100%|██████████| 133/133 [00:01<00:00, 108.09it/s]



Train set: Average loss: 0.0609
Test set: Average loss: 0.2342

Epoch: 25


100%|██████████| 133/133 [00:01<00:00, 92.03it/s]



Train set: Average loss: 0.0573
Test set: Average loss: 0.2386

Epoch: 26


100%|██████████| 133/133 [00:01<00:00, 118.38it/s]



Train set: Average loss: 0.0606
Test set: Average loss: 0.2568

Epoch: 27


100%|██████████| 133/133 [00:00<00:00, 301.50it/s]



Train set: Average loss: 0.0546
Test set: Average loss: 0.2468

Epoch: 28


100%|██████████| 133/133 [00:00<00:00, 302.27it/s]



Train set: Average loss: 0.0572
Test set: Average loss: 0.2412

Epoch: 29


100%|██████████| 133/133 [00:00<00:00, 318.46it/s]



Train set: Average loss: 0.0500
Test set: Average loss: 0.2591

Epoch: 30


100%|██████████| 133/133 [00:00<00:00, 306.35it/s]



Train set: Average loss: 0.0523
Test set: Average loss: 0.2457

Epoch: 31


100%|██████████| 133/133 [00:00<00:00, 308.35it/s]



Train set: Average loss: 0.0465
Test set: Average loss: 0.2408

Epoch: 32


100%|██████████| 133/133 [00:00<00:00, 274.12it/s]



Train set: Average loss: 0.0482
Test set: Average loss: 0.2275

Epoch: 33


100%|██████████| 133/133 [00:00<00:00, 305.89it/s]



Train set: Average loss: 0.0477
Test set: Average loss: 0.2385

Epoch: 34


100%|██████████| 133/133 [00:00<00:00, 292.62it/s]



Train set: Average loss: 0.0447
Test set: Average loss: 0.2279

Epoch: 35


100%|██████████| 133/133 [00:00<00:00, 300.84it/s]



Train set: Average loss: 0.0437
Test set: Average loss: 0.2597

Epoch: 36


100%|██████████| 133/133 [00:00<00:00, 301.55it/s]



Train set: Average loss: 0.0435
Test set: Average loss: 0.2467

Epoch: 37


100%|██████████| 133/133 [00:00<00:00, 298.95it/s]



Train set: Average loss: 0.0414
Test set: Average loss: 0.2397

Epoch: 38


100%|██████████| 133/133 [00:00<00:00, 284.42it/s]



Train set: Average loss: 0.0405
Test set: Average loss: 0.2515

Epoch: 39


100%|██████████| 133/133 [00:00<00:00, 288.01it/s]



Train set: Average loss: 0.0377
Test set: Average loss: 0.2574

Epoch: 40


100%|██████████| 133/133 [00:00<00:00, 295.14it/s]



Train set: Average loss: 0.0386
Test set: Average loss: 0.2460

Epoch: 41


100%|██████████| 133/133 [00:00<00:00, 290.88it/s]



Train set: Average loss: 0.0427
Test set: Average loss: 0.2568

Epoch: 42


100%|██████████| 133/133 [00:00<00:00, 297.89it/s]



Train set: Average loss: 0.0417
Test set: Average loss: 0.2573

Epoch: 43


100%|██████████| 133/133 [00:00<00:00, 302.81it/s]



Train set: Average loss: 0.0345
Test set: Average loss: 0.2591

Epoch: 44


100%|██████████| 133/133 [00:00<00:00, 304.70it/s]



Train set: Average loss: 0.0362
Test set: Average loss: 0.2628

Epoch: 45


100%|██████████| 133/133 [00:00<00:00, 203.09it/s]



Train set: Average loss: 0.0340
Test set: Average loss: 0.2499

Epoch: 46


100%|██████████| 133/133 [00:00<00:00, 220.17it/s]



Train set: Average loss: 0.0310
Test set: Average loss: 0.2755

Epoch: 47


100%|██████████| 133/133 [00:00<00:00, 194.25it/s]



Train set: Average loss: 0.0297
Test set: Average loss: 0.2626

Epoch: 48


100%|██████████| 133/133 [00:00<00:00, 183.41it/s]



Train set: Average loss: 0.0317
Test set: Average loss: 0.2441

Epoch: 49


100%|██████████| 133/133 [00:00<00:00, 282.22it/s]



Train set: Average loss: 0.0331
Test set: Average loss: 0.2659

Epoch: 50


100%|██████████| 133/133 [00:00<00:00, 281.73it/s]



Train set: Average loss: 0.0316
Test set: Average loss: 0.2589

Epoch: 51


100%|██████████| 133/133 [00:00<00:00, 295.82it/s]



Train set: Average loss: 0.0333
Test set: Average loss: 0.2564

Epoch: 52


100%|██████████| 133/133 [00:00<00:00, 306.63it/s]



Train set: Average loss: 0.0339
Test set: Average loss: 0.2672

Epoch: 53


100%|██████████| 133/133 [00:00<00:00, 317.98it/s]



Train set: Average loss: 0.0289
Test set: Average loss: 0.2624

Epoch: 54


100%|██████████| 133/133 [00:00<00:00, 293.30it/s]



Train set: Average loss: 0.0283
Test set: Average loss: 0.2629

Epoch: 55


100%|██████████| 133/133 [00:00<00:00, 302.45it/s]



Train set: Average loss: 0.0294
Test set: Average loss: 0.2769

Epoch: 56


100%|██████████| 133/133 [00:00<00:00, 290.69it/s]



Train set: Average loss: 0.0283
Test set: Average loss: 0.2480

Epoch: 57


100%|██████████| 133/133 [00:00<00:00, 300.67it/s]



Train set: Average loss: 0.0286
Test set: Average loss: 0.2659

Epoch: 58


100%|██████████| 133/133 [00:00<00:00, 275.29it/s]



Train set: Average loss: 0.0319
Test set: Average loss: 0.2863

Epoch: 59


100%|██████████| 133/133 [00:00<00:00, 292.81it/s]



Train set: Average loss: 0.0278
Test set: Average loss: 0.2789

Epoch: 60


100%|██████████| 133/133 [00:00<00:00, 307.44it/s]


Train set: Average loss: 0.0341
Test set: Average loss: 0.2497
Training is ended!

BEST EPOCH:  1 with Loss:  0.17096189426456712


Вывод:

32 нейрона:
BEST EPOCH:  42 with Loss:  0.1681321774548185

64 нейрона:
BEST EPOCH:  3 with Loss:  0.17034084483231124

128 нейронов:
BEST EPOCH:  2 with Loss:  0.16978198520100207

256 нейронов:
BEST EPOCH:  1 with Loss:  0.17096189426456712

Лучший результат у 32 нейронов, но в целом нет сильных отличий между количеством нейронов. При этом можно отметить скорость обучения - 32 нейрона лучше всего показали себя только на 42 эпохе, а 64/128/256 на 3/2/1 эпохе, что является показателем недо/переобучения